### ISSUE TITLE
Support futures runtimes other than Tokio

### ISSUE THREAD

**Issue by xmclark** [hyperium/tonic#152]
Tonic seems to have made a large investment on Hyper and Tokio ecoystem. This had the side effect of making other runtimes not work well with tonic.

A basic demo of this can be created using on any of the tonic examples:
1. Add `async-std = { version = "1.1.0", features = ["attributes"]}` to the dependencies list
2. Replace `#[tokio::main]` with `#[async_std::main]`
3. Run `cargo run --example helloworld-client` 

An error message like this will appear:
`Error: Error(Client, Error(Connect, Custom { kind: Other, error: "no current reactor" }))`

So my open question: How can tonic better support using other non-tokio runtimes?

**Comment by LucioFranco** [comment]
@xmclark hi! Tonic is still at an alpha release (soon to be 0.1) so its still early in the process. There are a few things that tie us to tokio/hyper, mainly the transport module which is behind a feature flag. The real abstracion for tonic is around `tower-service` and `http-body` crates which both do not tie to tokio/hyper. So in theory, if you have another http/2 implenetation you can just implement those abstractions and it will work.

The current reason why tonic is very tied to hyper/tokio is that the goal of this library was to provide a good out of box experience. This means we should try to optimize for users that may not have worked with futures/runtimes or even rust before. To achieve this there needs to be some form of tight coupling which is shown in the transport feature/module.

As for your issue directly, it looks like `async-std` doesn't provide the proper hooks to setup the tokio reactor/timer. So that is the error you are getting since the transport ties directly to tokio/hyper.

As far as I know there is no other http/2 implementation in rust besides hyper/h2 but if there were it is possible to swap them out which means runtime support as well.

As for now, I don't think there is much tonic can expose better to provide better interop. The transport module needs to be tightly coupled to support its feature set as of now. I think once we get closer to a more complete `AsyncRead/AsyncWrite` trait and tokio/hyper hit non alpha releases it will make sense.

That said, if you can get hyper to work with `async-std` it should be a very simple drop in. The transport module just provides more features than just hyper.

It may be worthwhile to better explain the higher-level abstractions that tonic provides.

I'd be happy to explore more and I am happy to review PRs that might be able to easily provide this ability but I don't quite see the easy path yet that wouldn't be very intrusive.

TL;DR: Tonic provides abstractions around `tower-service` and `http-body` that do not tie to tokio nor hyper, but the `transport` module/feature needs to be tightly coupled to provide its current feature set.

**Comment by Sherlock-Holo** [comment]
I try to replace tonic `Channel` with my custom client, it built with `async-std`, however when I run, it still panic because the whole program still need tokio runtime. I dig codes and find maybe `h2` crate depends on tokio and we can't replace `h2` in tonic like replacing `Channel`

**Comment by LucioFranco** [comment]
`h2` only needs tokio because it uses tokio's io types and sync types. Otherwise, it should work on any other executor if you compat the traits correctly.

**Comment by iho** [comment]
Do you have any progress with it?

**Comment by onsails** [comment]
Perhaps [async-compat](https://github.com/stjepang/async-compat) could help.

**Comment by LucioFranco** [comment]
There are likely some hidden tokio spawns in this crate that will need to be refactored out. I don't currently have the time but happy to help mentor someone through it.

**Comment by iho** [comment]
@LucioFranco what to do if that hidden tokio spawns inside dependencies? 
For example, I found that spawn https://github.com/hyperium/hyper/blob/master/src/client/connect/dns.rs#L120
How to fix it? Do I need to fork hyper? How to pass runtime instance to that place? 
 
Sorry for the dumb questions. I want to make it work.

**Comment by pruthvikar** [comment]
A [comment](https://github.com/hyperium/hyper/issues/2156#issuecomment-603414852) from @seanmonstar on providing your own runtime

> As stated, by default hyper assumes the Tokio runtime is used. You can disable the default features, and provide your own [connector](https://docs.rs/hyper/0.13.*/hyper/client/struct.Builder.html#method.build) for the Client.

**Comment by zhuxiujia** [comment]
async_std  runtime(Tokio dependencies are already built in)  
Conditional compilation is already supported  tokio02  and tokio03.

**Comment by davidpdrsn** [comment]
@LucioFranco Is this still something you think we should support? My initial impression is that tonic is pretty tightly coupled to tokio through hyper and tower. I guess once the transport stuff is extracted it might be possible 🤷

**Comment by LucioFranco** [comment]
Technically, other runtimes are supported but I don't think we need to go out of our way right now to support other runtimes out of the box. The maintenance cost is too high and I personally don't have time.

**Comment by xavier83** [comment]
I'm new here. @davidpdrsn  Is tonic now capable or running under async-std runtime? If so is there any short example for it. Thanks.

**Comment by davidpdrsn** [comment]
@xavier83 no tonic is still tokio only. I'm afraid are currently no plans to support other runtimes.
